### PR TITLE
Merge consecutive iteration loops

### PR DESCRIPTION
**Pull request by drewnoakes**
CPU profiling shows about 50% of `DependenciesSnapshot.FromChanges` time was spent in `DuplicatedDependenciesSnapshotFilter.BeforeAddOrUpdate`, mostly enumerating immutable dictionaries.

On loading `Roslyn.sln` <sup>1</sup> we spent 4% CPU time producing dependencies snapshots, half of which was in this method:

image

The normal case here is that no duplicate caption is found. This change halves the number of iterations required.

I'm looking at separating handling of top-level and non-top-level nodes across the dependencies node, at which point the number of items being iterated will drop considerably, along with that CPU percentage. In the meantime, this is an easy win.

<sup>1</sup> Based on a 30 second PerfView capture started when the "Preparing projects" dialog disappears.

### CODE REVIEW

**Comment by davkean** [comment]
For the future, you should be able to scope to particular events to get more concrete numbers, for example, I'd consider; 

Start: Microsoft-VisualStudio-Common/vs_solution_perf_load/Stop
Stop: https://github.com/dotnet/project-system/blob/d9f0d1d40b81fe3f50317a758b3474426c69cb63/src/Microsoft.VisualStudio.ProjectSystem.Managed/Telemetry/TelemetryEventName.cs#L35

File: src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/Tree/Dependencies/Snapshot/Filters/DuplicatedDependenciesSnapshotFilter.cs [hunk 1]
```diff
@@ -30,52 +30,59 @@ internal sealed class DuplicatedDependenciesSnapshotFilter : DependenciesSnapsho
...
-                    if (x.TopLevel
-                         && !x.Id.Equals(dependency.Id, StringComparison.OrdinalIgnoreCase)
-                         && StringComparers.DependencyProviderTypes.Equals(x.ProviderType, dependency.ProviderType)
-                         && x.Caption.StartsWith(dependency.Caption, StringComparison.OrdinalIgnoreCase)
-                         && x.Caption.Length >= adjustedLength
-                         && string.Compare(x.Caption, adjustedLength, x.OriginalItemSpec, 0, x.OriginalItemSpec.Length, StringComparison.OrdinalIgnoreCase) == 0)
+                    if (other.Caption.Length == dependency.Caption.Length)
                     {
+                        // Exact match.
+                        System.Diagnostics.Debug.Assert(matchingDependency == null, "matchingDependency should be null");
```
Comment (davkean) [inline src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/Tree/Dependencies/Snapshot/Filters/DuplicatedDependenciesSnapshotFilter.cs]: Assumes.Null.

File: src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/Tree/Dependencies/Snapshot/Filters/DuplicatedDependenciesSnapshotFilter.cs [hunk 2]
```diff
@@ -30,52 +30,59 @@ internal sealed class DuplicatedDependenciesSnapshotFilter : DependenciesSnapsho
...
+                        matchingDependency = other;
                         shouldApplyAlias = true;
                         break;
                     }
+
+                    // Prefix matches.
+                    // Check whether we have a match of form "Caption (ItemSpec)".
+
+                    string itemSpec = other.OriginalItemSpec;
+                    int expectedItemSpecIndex = dependency.Caption.Length + 2;
```
Comment (davkean) [inline src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/Tree/Dependencies/Snapshot/Filters/DuplicatedDependenciesSnapshotFilter.cs]: I found the previous way this was adding this more obvious; can you put this `// Account for " ("` or somethign after it?

File: src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/Tree/Dependencies/Snapshot/Filters/DuplicatedDependenciesSnapshotFilter.cs (hunk 1 above)
Comment (drewnoakes) [reply src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/Tree/Dependencies/Snapshot/Filters/DuplicatedDependenciesSnapshotFilter.cs]: `Assert` is conditional so wouldn't be in release code.

It's impossible for this value to be anything other than `null`, given the `break`. I'll just remove the assert.

File: src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/Tree/Dependencies/Snapshot/Filters/DuplicatedDependenciesSnapshotFilter.cs (hunk 2 above)
Comment (drewnoakes) [reply src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/Tree/Dependencies/Snapshot/Filters/DuplicatedDependenciesSnapshotFilter.cs]: Comment added.

It'd be nice if `" (".Length` was reduced to a constant by the compiler.

**Pull request merged**